python/dgl/convert.py

```diff
@@ -1643,8 +1643,107 @@ def _batcher(lst):
     return g.to(device)
 
 
-def to_networkx(g, node_attrs=None, edge_attrs=None):
-    """Convert a homogeneous graph to a NetworkX graph and return.
+def _to_networkx_homogeneous(g, node_attrs, edge_attrs):
+    src, dst = g.edges()
+    src = F.asnumpy(src)
+    dst = F.asnumpy(dst)
+    # xiangsx: Always treat graph as multigraph
+    nx_graph = nx.MultiDiGraph()
+    nx_graph.add_nodes_from(range(g.num_nodes()))
+    for eid, (u, v) in enumerate(zip(src, dst)):
+        nx_graph.add_edge(u, v, id=eid)
+
+    if node_attrs is not None:
+        for nid, attr in nx_graph.nodes(data=True):
+            feat_dict = g._get_n_repr(0, nid)
+            attr.update(
+                {key: F.squeeze(feat_dict[key], 0) for key in node_attrs}
+            )
+    if edge_attrs is not None:
+        for _, _, attr in nx_graph.edges(data=True):
+            eid = attr["id"]
+            feat_dict = g._get_e_repr(0, eid)
+            attr.update(
+                {key: F.squeeze(feat_dict[key], 0) for key in edge_attrs}
+            )
+    return nx_graph
+
+
+def _to_networkx_heterogeneous(
```

Comment:
Will it significantly simplify the implementation by leveraging https://docs.dgl.ai/generated/dgl.to_homogeneous.html#dgl.to_homogeneous? Intuitively, there might be a significant overlapping between the logic here and the logic in `to_homogeneous`.

Reply:
There is certainly some overlap between the two. I think the main difference is that `to_homogeneous` creates a new graph, which is not necessary while creating a NetworkX graph IMHO (although my first approach was to use _exactly_ `to_homogeneous` and adjust the indices in order to create the NetworkX graph). Also, as far as I can tell, `to_homogeneous` is not being explicitly tested.

Would you like me to look for ways to extract common code?

Reply:
Then, will it suffice to simply call `to_homogeneous`, make some adjustments, and then call the `to_networkx` implementation for homogeneous graphs? This might significantly simplify your implementation. Over the years, an important lesson I've learnt is that we should try avoiding adding new code unless it's really necessary. In particular, there can be nasty maintenance efforts.

Reply:
It would do, at the expense of creating an intermediate graph object. Do you consider this an acceptable compromise?

Reply:
Ok then, I'll update the code to make use of `to_homogeneous` as much as possible. 

I think it would make sense to add some note in the docstring saying that the conversion creates a homogeneous graph internally and, as such, incurs extra memory cost. What do you think?

Reply:
I don't think you need to add that kind of note. Many of DGL's modules/utilities can incur extra memory cost. Creating a NetworkX object may also incur extra memory cost.

Reply:
Fine by me. Can you also confirm/reject the [assumption](https://github.com/dmlc/dgl/pull/5726#discussion_r1206340811) on etypes?

Reply:
Interestingly, `to_homogeneous` does not seem to work right away with `ndata` and `edata`. I'll investigate.

Reply:
From further investigation, I've found that `to_homogeneous` does not work in two cases, namely:
- when a node/edge feature is not assigned to all node/edge types;
- when its shape is different across node/edge types. 

A minimal example:

```
g = dgl.heterograph(
    {
        ("user", "follows", "user"): (
            torch.tensor([0, 1, 1, 2, 2, 2]),
            torch.tensor([0, 0, 1, 1, 2, 2]),
        ),
        ("user", "plays", "game"): (
            torch.tensor([0, 1]),
            torch.tensor([0, 1]),
        ),
    }
)
g.nodes["user"].data["h"] = torch.ones(3, 1)
g.edges["plays"].data["h"] = torch.ones(2, 1)
```

Both `dgl.to_homogeneous(g, ndata=["h"])` and `dgl.to_homogeneous(g, edata=["h"])` fail with `KeyError: 'h'` (the issue is in the `combine_frames` function of the `heterograph` module).

Adding feature `h` on both the `game` node type and the `follows` edge type with a different shape (second dimension of the `torch.ones` vector being 2 instead of 1) fails as well in the same method. For example, the following:

```
g.nodes["game"].data["h"] = torch.ones(2, 2)
dgl.to_homogeneous(g, ndata=["h"])
```

fails with `DGLError: Cannot concatenate column h with shape Scheme(shape=(1,), dtype=torch.float32) and shape Scheme(shape=(2,), dtype=torch.float32)`, and the same happens with the following:

```
g.edges["follows"].data["h"] = torch.ones(6, 2)
dgl.to_homogeneous(g, edata=["h"])
```

Therefore, `to_homogeneous` cannot be used as a direct replacement if this is the desired behaviour.

Am I correct in assuming that both these scenarios are valid use cases?

Reply:
I'm fine either way. Will you encounter any of these cases in your scenario?

Option one, we can choose not to support these two cases for this PR and wait to see if there will be many feature requests. We might want to improve the warning messages to communicate better.

Option two, we support these two cases. For this route, we can use `to_homogeneous` to renumber the node/edge IDs and process the node/edge feature separately.

Reply:
I have no direct use case at the moment, but I think they are reasonable scenarios as typically different node/edge types have different attributes assigned. 

Unless you specifically want to rule these out because of different assumptions (perhaps related to uniformity?), then option two seems more natural.

Reply:
I've refactored the internal function to use `to_homogeneous` and process ndata/edata separately, with a comment to explain the choice.